File: tests/metrics_collector.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <time.h>
#include <sys/resource.h>
#include <sys/time.h>
#include <pthread.h>
#include <signal.h>
#include <dirent.h>

#define MAX_LINE_LENGTH 512

typedef struct {
    char output_file[256];
    pid_t target_pid;
    int sample_interval;
    volatile int running;
} MetricsCollector;

typedef struct {
    time_t timestamp;
    // CPU метрики
    double cpu_percent;
    unsigned long cpu_time_user;
    unsigned long cpu_time_system;
    // Memory метрики
    unsigned long mem_rss_kb;
    unsigned long mem_vsz_kb;
    // IO метрики
    unsigned long io_read_bytes;
    unsigned long io_write_bytes;
    unsigned long io_read_ops;
    unsigned long io_write_ops;
    // Thread метрики
    int num_threads;
    // File descriptor метрики
    int num_fds;
    // Context switch метрики
    unsigned long ctx_switches_voluntary;
    unsigned long ctx_switches_involuntary;
    // Page fault метрики
    unsigned long page_faults_minor;
    unsigned long page_faults_major;
} ProcessMetrics;

static MetricsCollector collector;

// Получение метрик CPU и памяти из /proc/[pid]/stat
static int get_cpu_mem_metrics(pid_t pid, ProcessMetrics *metrics) {
    char path[256];
    snprintf(path, sizeof(path), "/proc/%d/stat", pid);
    
    FILE *f = fopen(path, "r");
    if (!f) return -1;
    
    char line[MAX_LINE_LENGTH];
    if (!fgets(line, sizeof(line), f)) {
        fclose(f);
        return -1;
    }
    fclose(f);
    
    // Парсинг /proc/[pid]/stat
    char *ptr = strrchr(line, ')');
    if (!ptr) return -1;
    ptr += 2;
    
    unsigned long utime, stime, vsize, rss, minflt, majflt;
    int num_threads;
    
    // Формат: state ppid pgrp session tty_nr tpgid flags minflt cminflt majflt cmajflt utime stime...
    sscanf(ptr, "%*c %*d %*d %*d %*d %*d %*u %lu %*u %lu %*u %lu %lu %*d %*d %*d %*d %d %*d %*u %lu %ld",
           &minflt, &majflt, &utime, &stime, &num_threads, &vsize, &rss);
    
    metrics->cpu_time_user = utime;
    metrics->cpu_time_system = stime;
    metrics->mem_vsz_kb = vsize / 1024;
    metrics->mem_rss_kb = rss * (sysconf(_SC_PAGESIZE) / 1024);
    metrics->num_threads = num_threads;
    metrics->page_faults_minor = minflt;
    metrics->page_faults_major = majflt;
    
    return 0;
}

// Вычисление процента CPU
static double calculate_cpu_percent(unsigned long prev_total, unsigned long curr_total, 
                                    double elapsed_sec) {
    unsigned long delta = curr_total - prev_total;
    long hz = sysconf(_SC_CLK_TCK);
    double cpu_percent = (delta * 100.0) / (hz * elapsed_sec);
    return cpu_percent;
}

// Получение IO метрик из /proc/[pid]/io
static int get_io_metrics(pid_t pid, ProcessMetrics *metrics) {
    char path[256];
    snprintf(path, sizeof(path), "/proc/%d/io", pid);
    
    FILE *f = fopen(path, "r");
    if (!f) return -1;
    
    char line[MAX_LINE_LENGTH];
    while (fgets(line, sizeof(line), f)) {
        if (strncmp(line, "read_bytes:", 11) == 0) {
            sscanf(line + 12, "%lu", &metrics->io_read_bytes);
        } else if (strncmp(line, "write_bytes:", 12) == 0) {
            sscanf(line + 13, "%lu", &metrics->io_write_bytes);
        } else if (strncmp(line, "syscr:", 6) == 0) {
            sscanf(line + 7, "%lu", &metrics->io_read_ops);
        } else if (strncmp(line, "syscw:", 6) == 0) {
            sscanf(line + 7, "%lu", &metrics->io_write_ops);
        }
    }
    fclose(f);
    return 0;
}

// Получение context switches из /proc/[pid]/status
static int get_context_switches(pid_t pid, ProcessMetrics *metrics) {
    char path[256];
    snprintf(path, sizeof(path), "/proc/%d/status", pid);
    
    FILE *f = fopen(path, "r");
    if (!f) return -1;
    
    char line[MAX_LINE_LENGTH];
    while (fgets(line, sizeof(line), f)) {
        if (strncmp(line, "voluntary_ctxt_switches:", 24) == 0) {
            sscanf(line + 25, "%lu", &metrics->ctx_switches_voluntary);
        } else if (strncmp(line, "nonvoluntary_ctxt_switches:", 27) == 0) {
            sscanf(line + 28, "%lu", &metrics->ctx_switches_involuntary);
        }
    }
    fclose(f);
    return 0;
}

// Подсчет открытых файловых дескрипторов через readdir
static int count_fds(pid_t pid) {
    char path[256];
    snprintf(path, sizeof(path), "/proc/%d/fd", pid);
    
    DIR *dir = opendir(path);
    if (!dir) return 0;
    
    int count = 0;
    struct dirent *entry;
    while ((entry = readdir(dir)) != NULL) {
        if (entry->d_name[0] != '.') {
            count++;
        }
    }
    closedir(dir);
    return count;
}

// Сбор всех метрик
static int collect_metrics(pid_t pid, ProcessMetrics *metrics, 
                          ProcessMetrics *prev_metrics, double elapsed_sec) {
    metrics->timestamp = time(NULL);
    
    if (get_cpu_mem_metrics(pid, metrics) != 0) return -1;
    if (get_io_metrics(pid, metrics) != 0) return -1;
    if (get_context_switches(pid, metrics) != 0) return -1;
    
    metrics->num_fds = count_fds(pid);
    
    // Вычисление процента CPU
    if (prev_metrics && elapsed_sec > 0) {
        unsigned long prev_total = prev_metrics->cpu_time_user + prev_metrics->cpu_time_system;
        unsigned long curr_total = metrics->cpu_time_user + metrics->cpu_time_system;
        metrics->cpu_percent = calculate_cpu_percent(prev_total, curr_total, elapsed_sec);
    } else {
        metrics->cpu_percent = 0.0;
    }
    
    return 0;
}

// Запись метрик в CSV файл
static void write_metrics_to_csv(FILE *f, ProcessMetrics *metrics) {
    fprintf(f, "%ld,%.2f,%lu,%lu,%lu,%lu,%lu,%lu,%lu,%lu,%d,%d,%lu,%lu,%lu,%lu\n",
            metrics->timestamp,
            metrics->cpu_percent,
            metrics->cpu_time_user,
            metrics->cpu_time_system,
            metrics->mem_rss_kb,
            metrics->mem_vsz_kb,
            metrics->io_read_bytes,
            metrics->io_write_bytes,
            metrics->io_read_ops,
            metrics->io_write_ops,
            metrics->num_threads,
            metrics->num_fds,
            metrics->ctx_switches_voluntary,
            metrics->ctx_switches_involuntary,
            metrics->page_faults_minor,
            metrics->page_faults_major);
    fflush(f);
}

// Обработчик сигналов для graceful shutdown
static void signal_handler(int _) {
    (void)_;
    collector.running = 0;
}

// Основной цикл сбора метрик
static void *collector_thread(void *_) {
    (void)_;
    FILE *output = fopen(collector.output_file, "w");
    if (!output) {
        fprintf(stderr, "Failed to open output file: %s\n", collector.output_file);
        return NULL;
    }
    
    // Записываем заголовок CSV
    fprintf(output, "timestamp,cpu_percent,cpu_user,cpu_system,mem_rss_kb,mem_vsz_kb,"
                   "io_read_bytes,io_write_bytes,io_read_ops,io_write_ops,threads,fds,"
                   "ctx_switches_voluntary,ctx_switches_involuntary,page_faults_minor,page_faults_major\n");
    fflush(output);
    
    ProcessMetrics current, previous;
    memset(&previous, 0, sizeof(previous));
    memset(&current, 0, sizeof(current));
    
    struct timespec prev_time, curr_time;
    clock_gettime(CLOCK_MONOTONIC, &prev_time);
    
    // Делаем первый замер для инициализации
    if (collect_metrics(collector.target_pid, &previous, NULL, 0) != 0) {
        fprintf(stderr, "Failed to collect initial metrics\n");
        fclose(output);
        return NULL;
    }
    
    while (collector.running) {
        sleep(collector.sample_interval);
        
        clock_gettime(CLOCK_MONOTONIC, &curr_time);
        double elapsed = (curr_time.tv_sec - prev_time.tv_sec) + 
                        (curr_time.tv_nsec - prev_time.tv_nsec) / 1e9;
        
        if (collect_metrics(collector.target_pid, &current, &previous, elapsed) == 0) {
            write_metrics_to_csv(output, &current);
            previous = current;
            prev_time = curr_time;
        } else {
            fprintf(stderr, "Failed to collect metrics (process may have died)\n");
            break;
        }
    }
    
    fclose(output);
    return NULL;
}

int main(int argc, char *argv[]) {
    if (argc < 3) {
        fprintf(stderr, "Usage: %s <pid> <output_csv_file> [sample_interval_sec]\n", argv[0]);
        fprintf(stderr, "Example: %s 12345 metrics.csv 1\n", argv[0]);
        return 1;
    }
    
    collector.target_pid = atoi(argv[1]);
    strncpy(collector.output_file, argv[2], sizeof(collector.output_file) - 1);
    collector.sample_interval = (argc >= 4) ? atoi(argv[3]) : 1;
    collector.running = 1;
    
    if (collector.sample_interval < 1) {
        collector.sample_interval = 1;
    }
    
    // Проверяем существование процесса
    if (kill(collector.target_pid, 0) != 0) {
        fprintf(stderr, "Process %d does not exist or not accessible\n", collector.target_pid);
        return 1;
    }
    
    // Устанавливаем обработчики сигналов
    signal(SIGINT, signal_handler);
    signal(SIGTERM, signal_handler);
    
    printf("Collecting metrics for PID %d\n", collector.target_pid);
    printf("Output file: %s\n", collector.output_file);
    printf("Sample interval: %d second(s)\n", collector.sample_interval);
    printf("Press Ctrl+C to stop\n\n");
    
    pthread_t thread;
    pthread_create(&thread, NULL, collector_thread, NULL);
    pthread_join(thread, NULL);
    
    printf("\nMetrics collection stopped. Results saved to %s\n", collector.output_file);
    return 0;
}

File: tests/metrics_collector.go
package main

import (
	"bufio"
	"context"
	"fmt"
	"os"
	"os/signal"
	"strconv"
	"strings"
	"sync"
	"syscall"
	"time"
)

// ProcessMetrics holds collected metrics for a process
type ProcessMetrics struct {
	Timestamp              time.Time
	CPUPercent             float64
	CPUTimeUser            uint64
	CPUTimeSystem          uint64
	MemRSSKB               uint64
	MemVSZKB               uint64
	IOReadBytes            uint64
	IOWriteBytes           uint64
	IOReadOps              uint64
	IOWriteOps             uint64
	NumThreads             int
	NumFDs                 int
	CtxSwitchesVoluntary   uint64
	CtxSwitchesInvoluntary uint64
	PageFaultsMinor        uint64
	PageFaultsMajor        uint64
}

// MetricsCollector holds collector state
type MetricsCollector struct {
	PID            int
	OutputFile     string
	SampleInterval time.Duration
	running        bool
	mu             sync.Mutex
}

func main() {
	if len(os.Args) < 3 {
		fmt.Fprintf(os.Stderr, "Usage: %s <pid> <output_csv> [sample_interval_sec]\n", os.Args[0])
		fmt.Fprintf(os.Stderr, "Example: %s 12345 metrics.csv 1\n", os.Args[0])
		os.Exit(1)
	}

	pid, err := strconv.Atoi(os.Args[1])
	if err != nil {
		fmt.Fprintf(os.Stderr, "Invalid PID: %v\n", err)
		os.Exit(1)
	}

	outputFile := os.Args[2]
	sampleInterval := 1 * time.Second
	if len(os.Args) >= 4 {
		seconds, err := strconv.Atoi(os.Args[3])
		if err != nil || seconds < 1 {
			seconds = 1
		}
		sampleInterval = time.Duration(seconds) * time.Second
	}

	// Check if process exists
	if !processExists(pid) {
		fmt.Fprintf(os.Stderr, "Process %d does not exist or not accessible\n", pid)
		os.Exit(1)
	}

	collector := &MetricsCollector{
		PID:            pid,
		OutputFile:     outputFile,
		SampleInterval: sampleInterval,
		running:        true,
	}

	fmt.Printf("Collecting metrics for PID %d\n", pid)
	fmt.Printf("Output file: %s\n", outputFile)
	fmt.Printf("Sample interval: %s\n", sampleInterval)
	fmt.Println("Press Ctrl+C to stop")

	if err := collector.Run(); err != nil {
		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
		os.Exit(1)
	}

	fmt.Printf("\nMetrics collection stopped. Results saved to %s\n", outputFile)
}

// processExists checks if a process with given PID exists
func processExists(pid int) bool {
	if pid <= 0 {
		return false
	}
	process, err := os.FindProcess(pid)
	if err != nil {
		return false
	}
	err = process.Signal(syscall.Signal(0))
	return err == nil
}

// Run starts the metrics collection loop
func (c *MetricsCollector) Run() error {
	// Create context for graceful shutdown
	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	// Handle signals
	sigChan := make(chan os.Signal, 1)
	signal.Notify(sigChan, os.Interrupt, syscall.SIGTERM)
	go func() {
		<-sigChan
		c.mu.Lock()
		c.running = false
		c.mu.Unlock()
		cancel()
	}()

	// Open output file
	f, err := os.Create(c.OutputFile)
	if err != nil {
		return fmt.Errorf("failed to create output file: %w", err)
	}
	defer f.Close()

	writer := bufio.NewWriter(f)
	defer writer.Flush()

	// Write CSV header
	header := "timestamp,cpu_percent,cpu_user,cpu_system,mem_rss_kb,mem_vsz_kb," +
		"io_read_bytes,io_write_bytes,io_read_ops,io_write_ops,threads,fds," +
		"ctx_switches_voluntary,ctx_switches_involuntary,page_faults_minor,page_faults_major\n"
	if _, err := writer.WriteString(header); err != nil {
		return fmt.Errorf("failed to write header: %w", err)
	}

	// Initial metrics collection for baseline
	var previous, current ProcessMetrics
	if err := c.collectMetrics(&previous); err != nil {
		return fmt.Errorf("failed to collect initial metrics: %w", err)
	}

	prevTime := time.Now()

	// Main collection loop
	ticker := time.NewTicker(c.SampleInterval)
	defer ticker.Stop()

	for {
		c.mu.Lock()
		running := c.running
		c.mu.Unlock()

		if !running {
			break
		}

		select {
		case <-ticker.C:
			now := time.Now()
			elapsed := now.Sub(prevTime).Seconds()

			if err := c.collectMetrics(&current); err != nil {
				fmt.Fprintf(os.Stderr, "Failed to collect metrics: %v\n", err)
				continue
			}

			// Calculate CPU percent
			if elapsed > 0 {
				prevTotal := previous.CPUTimeUser + previous.CPUTimeSystem
				currTotal := current.CPUTimeUser + current.CPUTimeSystem
				current.CPUPercent = calculateCPUPercent(prevTotal, currTotal, elapsed)
			} else {
				current.CPUPercent = 0
			}

			// Write metrics to CSV
			if err := c.writeMetrics(writer, &current); err != nil {
				fmt.Fprintf(os.Stderr, "Failed to write metrics: %v\n", err)
			}

			previous = current
			prevTime = now

		case <-ctx.Done():
			return nil
		}
	}

	return nil
}

// collectMetrics gathers all metrics for the target PID
func (c *MetricsCollector) collectMetrics(metrics *ProcessMetrics) error {
	metrics.Timestamp = time.Now()

	// Get CPU and memory metrics from /proc/[pid]/stat
	if err := c.getCPUMemMetrics(metrics); err != nil {
		return err
	}

	// Get IO metrics from /proc/[pid]/io
	if err := c.getIOMetrics(metrics); err != nil {
		return err
	}

	// Get context switches from /proc/[pid]/status
	if err := c.getContextSwitches(metrics); err != nil {
		return err
	}

	// Count file descriptors
	metrics.NumFDs = c.countFDs()

	return nil
}

// getCPUMemMetrics reads /proc/[pid]/stat
func (c *MetricsCollector) getCPUMemMetrics(metrics *ProcessMetrics) error {
	path := fmt.Sprintf("/proc/%d/stat", c.PID)
	data, err := os.ReadFile(path)
	if err != nil {
		return err
	}

	fields := strings.Fields(string(data))
	if len(fields) < 24 {
		return fmt.Errorf("invalid stat format")
	}

	// Parse fields from /proc/[pid]/stat
	// utime (field 14, index 13), stime (field 15, index 14)
	// minflt (field 10, index 9), majflt (field 12, index 11)
	// num_threads (field 20, index 19), vsize (field 23, index 22), rss (field 24, index 23)

	utime, _ := strconv.ParseUint(fields[13], 10, 64)
	stime, _ := strconv.ParseUint(fields[14], 10, 64)
	minflt, _ := strconv.ParseUint(fields[9], 10, 64)
	majflt, _ := strconv.ParseUint(fields[11], 10, 64)
	numThreads, _ := strconv.Atoi(fields[19])
	vsize, _ := strconv.ParseUint(fields[22], 10, 64)
	rss, _ := strconv.ParseUint(fields[23], 10, 64)

	metrics.CPUTimeUser = utime
	metrics.CPUTimeSystem = stime
	metrics.PageFaultsMinor = minflt
	metrics.PageFaultsMajor = majflt
	metrics.NumThreads = numThreads
	metrics.MemVSZKB = vsize / 1024
	metrics.MemRSSKB = rss * uint64(os.Getpagesize()/1024)

	return nil
}

// getIOMetrics reads /proc/[pid]/io
func (c *MetricsCollector) getIOMetrics(metrics *ProcessMetrics) error {
	path := fmt.Sprintf("/proc/%d/io", c.PID)
	f, err := os.Open(path)
	if err != nil {
		return err
	}
	defer f.Close()

	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		line := scanner.Text()
		parts := strings.SplitN(line, ":", 2)
		if len(parts) != 2 {
			continue
		}

		key := strings.TrimSpace(parts[0])
		valueStr := strings.TrimSpace(parts[1])
		value, _ := strconv.ParseUint(valueStr, 10, 64)

		switch key {
		case "read_bytes":
			metrics.IOReadBytes = value
		case "write_bytes":
			metrics.IOWriteBytes = value
		case "syscr":
			metrics.IOReadOps = value
		case "syscw":
			metrics.IOWriteOps = value
		}
	}

	return scanner.Err()
}

// getContextSwitches reads /proc/[pid]/status
func (c *MetricsCollector) getContextSwitches(metrics *ProcessMetrics) error {
	path := fmt.Sprintf("/proc/%d/status", c.PID)
	f, err := os.Open(path)
	if err != nil {
		return err
	}
	defer f.Close()

	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		line := scanner.Text()
		parts := strings.SplitN(line, ":", 2)
		if len(parts) != 2 {
			continue
		}

		key := strings.TrimSpace(parts[0])
		valueStr := strings.TrimSpace(parts[1])
		value, _ := strconv.ParseUint(valueStr, 10, 64)

		switch key {
		case "voluntary_ctxt_switches":
			metrics.CtxSwitchesVoluntary = value
		case "nonvoluntary_ctxt_switches":
			metrics.CtxSwitchesInvoluntary = value
		}
	}

	return scanner.Err()
}

// countFDs counts file descriptors in /proc/[pid]/fd
func (c *MetricsCollector) countFDs() int {
	path := fmt.Sprintf("/proc/%d/fd", c.PID)
	entries, err := os.ReadDir(path)
	if err != nil {
		return 0
	}

	count := 0
	for _, entry := range entries {
		if entry.Name()[0] != '.' {
			count++
		}
	}
	return count
}

// writeMetrics writes metrics to CSV
func (c *MetricsCollector) writeMetrics(writer *bufio.Writer, metrics *ProcessMetrics) error {
	timestamp := metrics.Timestamp.Unix()
	line := fmt.Sprintf("%d,%.2f,%d,%d,%d,%d,%d,%d,%d,%d,%d,%d,%d,%d,%d,%d,%d\n",
		timestamp,
		metrics.CPUPercent,
		metrics.CPUTimeUser,
		metrics.CPUTimeSystem,
		metrics.MemRSSKB,
		metrics.MemVSZKB,
		metrics.IOReadBytes,
		metrics.IOWriteBytes,
		metrics.IOReadOps,
		metrics.IOWriteOps,
		metrics.NumThreads,
		metrics.NumFDs,
		metrics.CtxSwitchesVoluntary,
		metrics.CtxSwitchesInvoluntary,
		metrics.PageFaultsMinor,
		metrics.PageFaultsMajor)

	_, err := writer.WriteString(line)
	if err != nil {
		return err
	}
	return writer.Flush()
}

// calculateCPUPercent calculates CPU percentage from delta
func calculateCPUPercent(prevTotal, currTotal uint64, elapsedSec float64) float64 {
	if elapsedSec <= 0 {
		return 0
	}
	delta := float64(currTotal - prevTotal)
	hz := float64(os.Getpagesize()) // Approximate clock ticks
	return (delta * 100.0) / (hz * elapsedSec)
}
